File: src/jimn/coordinates_hash.py
# vim : tabstop=4 expandtab shiftwidth=4 softtabstop=4
from jimn.precision import coordinate_key, displaced_coordinate_key, precision
from jimn.point import point
from jimn.debug import is_module_debugged


class coordinates_hash:
    def __init__(self, dimension, wanted_precision=precision):
        self.hashes = []
        self.precision = wanted_precision
        for d in range(dimension):
            self.hashes.append({})

    def hash_coordinate(self, index, c):
        key = coordinate_key(c, self.precision)
        displaced_key = displaced_coordinate_key(c)

        if key in self.hashes[index]:
            c = self.hashes[index][key]
        else:
            if displaced_key in self.hashes[index]:
                c = self.hashes[index][displaced_key]

        self.hashes[index][key] = c
        self.hashes[index][displaced_key] = c

        return c

    def hash_point(self, p):
        new_coordinates = []
        for i, c in enumerate(p.get_coordinates()):
            new_c = self.hash_coordinate(i, c)
            new_coordinates.append(new_c)
            if __debug__:
                if is_module_debugged(__name__):
                    print("coordinate {} hashes to {}".format(c, new_c))

        new_point = point(new_coordinates)
        return new_point

File: rust/jimn/examples/coordinates_hash.rs
extern crate jimn;
use jimn::utils::coordinates_hash::CoordinatesHash;
use jimn::point::Point;

fn main() {
    let points = [
        Point::new(0.123456789, 0.0),
        Point::new(0.123456788, 0.0),
        Point::new(0.123456778, 0.0),
        Point::new(0.123456678, 0.0),
        Point::new(0.123455678, 0.0),
        Point::new(0.123445678, 0.0),
        Point::new(0.123345678, 0.0),
        Point::new(0.122345678, 0.0),
        Point::new(0.112345678, 0.0),
        Point::new(0.012345678, 0.0),
    ];
    let mut hash = CoordinatesHash::new(2, 3);
    for point in &points {
        let hashed_point = hash.hash_point(point);
        println!("{:?} hashes to {:?}", point, hashed_point);
    }
}
